File: 0035-search-insert-position/0035-search-insert-position.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var searchInsert = function(nums, target) {
    if(nums == null || nums.length == 0) return 0;
		
    var n = nums.length;
    var l = 0;
    var r = n - 1;
    while(l < r){
        var m = l + Math.floor((r - l)/2);
        if(nums[m] == target) return m;
        else if(nums[m] > target) r = m; // right could be the result
        else l = m + 1; // m + 1 could be the result
    }

    // 1 element left at the end
    // post-processing
    return nums[l] < target ? l + 1: l;
};

File: 0035-search-insert-position/0035-search-insert-position.c
int searchInsert(int* nums, int numsSize, int target){
    int start = 0, end = numsSize - 1, mid;
    while(end >= start){
        mid = start + ((end - start) >> 1);
        if(target == nums[mid]){
            return mid;
        } else if (target > nums[mid]){
            start = mid + 1;
        } else {
            end = mid - 1;
        }
    }
    return nums[mid] > target ? mid : mid + 1;
}